pkg/epp/datalayer/metrics/client.go

```
@@ -0,0 +1,135 @@
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datalayer"
+)
+
+// Client is an interface for retrieving the metrics from an endpoint URL.
+type Client interface {
+	Get(ctx context.Context, target *url.URL, ep datalayer.Addressable) (PrometheusMetricMap, error)
+}
+
+// ClientFactory returns a Client suitable for an endpoint.
+// Implementations may return a new Client each time, or use a cached
+// copy for optimized retrieval.
+type ClientFactory interface {
+	GetClientForEndpoint(ep datalayer.Addressable) Client
+}
+
+// GetDefaultClientFactory returns a default implementation of the
+// ClientFactory, which caches and reuses client across calls.
+func GetDefaultClientFactory() ClientFactory {
+	return defaultClientFactory
+}
+
+// -- package implementations --
+var (
+	cleanupTick          = 30 * time.Second
+	maxIdleTime          = time.Minute
+	defaultClientFactory = newClientFactory()
+)
+
+type client struct {
+	cl       *http.Client
+	lastUsed time.Time
+}
+
+type clientmap struct {
+	cleanupOnce sync.Once
+	clients     sync.Map // map[string]*client (uses client IP as key)
+}
+
+func newClientFactory() *clientmap {
+	clm := &clientmap{}
+	clm.startCleanupGoroutine()
+	return clm
+}
+
+func (clm *clientmap) startCleanupGoroutine() {
+	clm.cleanupOnce.Do(func() {
+		go func() {
+			for {
+				time.Sleep(cleanupTick)
+				now := time.Now()
+
+				clm.clients.Range(func(key, value any) bool {
+					entry := value.(*client)
+					if now.Sub(entry.lastUsed) > maxIdleTime {
+						clm.clients.Delete(key)
+					}
+					return true
+				})
+			}
+		}()
+	})
+}
+
+func (clm *clientmap) GetClientForEndpoint(ep datalayer.Addressable) Client {
+	cl := newClient()
+	id := ep.GetIPAddress()
+
+	if value, loaded := clm.clients.LoadOrStore(id, cl); !loaded {
+		cl = value.(*client) // let it panic if need be
```

panic is a big no no.
always prefer to return an error.

---

agree, but "it should never panic" (c).
we store `*client` into it so nothing else should come out.

---

Why not just capture the second return val (a bool) and log if its hit? That ensures graceful handling & provides a paper trail to an on-caller/SRE for debugging.